pymeasure/adapters/adapter.py

```diff
@@ -34,44 +37,112 @@ class Adapter:
     This class should only be inherited from.
 
     :param preprocess_reply: optional callable used to preprocess strings
-        received from the instrument. The callable returns the processed string.
+        received from the instrument. The callable returns the processed
+        string. This argument is deprecated.
     :param kwargs: all other keyword arguments are ignored.
     """
 
     def __init__(self, preprocess_reply=None, **kwargs):
         self.preprocess_reply = preprocess_reply
         self.connection = None
+        self.log = logging.Logger("Adapter")
+        self.log.addHandler(logging.NullHandler())
+        if preprocess_reply is not None:
+            warn(
+                "Do not use `Adapter.values`, but `Instrument.values` instead. "
+                "Preprocess_reply is a values parameter.",
+                FutureWarning)
 
     def __del__(self):
         """Close connection upon garbage collection of the device"""
-        if self.connection is not None:
+        self.close()
+
+    def close(self):
+        """Close the connection."""
+        try:
             self.connection.close()
+        except AttributeError:
+            pass
 
-    def write(self, command):
-        """ Writes a command to the instrument
+    # Directly called methods. DO NOT OVERRIDE IN SUBCLASS!
```

Comment:
I'm wondering -- how do we catch/detect code in the wild (not in our codebase) that still overrides the "old" way, and clobbers our read/write/.. methods? Do you have an idea? A sentinel value set in our methods, that gets checked on `Adapter.__del__` and warns if not updated (so execution has never passed our read/write/...?

Reply:
Their code will work, but they will miss the major change: logging.
If they have their major logic in the adapter's values / aks methods, their code will be broken and they will see the changes in the changelog once they investigate. If they invoke via super, they will already get the warning.

I don't see the benefit (yet) of adding complexity in order to catch some obscure code which does not cause trouble (missing logging is no trouble, if you don't know that you can log).

Reply:
I agree that if it's only logging that goes missing, it's not a big deal. 
I'm more worrying about breaking user code, it could be unexpected and hard to find at first, especially if people are not careful about updating w/o reading changelogs. 
In any case, I'll make a prominent entry in the changelog.